File: Kotlin/BlackJack/src/BlackJack/entity/Calculator.kt
package BlackJack.entity

import BlackJack.type.CardType
import BlackJack.type.ResultType

object Calculator {

    fun evaluate(cards: ArrayList<Card>) : Int{

        var sum = 0
        cards.forEach { sum += it.value.number() }

        if(cards.any { it.value == CardType.ACE }) {

            var numberOfAces = cards.count { it.value == CardType.ACE }

            while(sum > 21 && numberOfAces > 0) {
                sum -= 10
                numberOfAces--
            }
        }
        return sum
    }

    fun calculateResult(player: Deck, bank: Deck) : ResultType {

        val playerValue = evaluate(player.cards)
        val bankValue = evaluate(bank.cards)

        if (playerValue == 21 && player.cards.count() == 2)
            if (player.cards.any { it.value == CardType.JACK } && player.cards.any { it.value == CardType.TEN }) return ResultType.WINBYJACK

        if (bankValue == 21 && bank.cards.count() == 2)
            if (bank.cards.any { it.value == CardType.JACK } && bank.cards.any { it.value == CardType.TEN }) return ResultType.LOSEBYJACK


        when {
            playerValue > 21 -> return ResultType.LOSE
            playerValue == 21 -> {
                return if (bankValue == 21) ResultType.TIE
                else ResultType.WIN
            }
            else -> return when {
                bankValue == 21 -> ResultType.LOSE
                bankValue > 21 -> ResultType.WIN
                else -> {
                    when {
                        bankValue < playerValue -> ResultType.WIN
                        bankValue == playerValue -> ResultType.TIE
                        else -> ResultType.LOSE
                    }
                }
            }
        }
    }


}

File: Lua/BlackJack/blackjack/entity/Calculator.lua

local ResultType = require "type.ResultType"

local Calculator = {}

function Calculator:getInstance()

    local evaluate = function(cards)
        local sum = 0

        for _, card in ipairs(cards) do
            sum = sum + card.number
        end

        local numberOfAces = 0

        for _, card in ipairs(cards) do
            if(card.value == "ACE")
            then
                numberOfAces = numberOfAces + 1
            end
        end

        while(sum > 21 and numberOfAces > 0) do
            sum = sum - 10  -- Making the value of ACE (default 11) to 1
            numberOfAces = numberOfAces - 1
        end

        return sum

    end

    local calculateResult = function(playerDeck, bankDeck)

        local playerValue = evaluate(playerDeck.cards)
        local bankValue = evaluate(bankDeck.cards)

        if(playerValue == 21 and #playerDeck.cards == 2) then
            if((playerDeck.cards[1].value == "JACK" or playerDeck.cards[2].value == "JACK") and
               (playerDeck.cards[1].value == "TEN" or playerDeck.cards[2].value == "TEN")) then
                return ResultType.WINBYJACK
            end
        end

        if(bankValue == 21 and #bankDeck.cards == 2) then
            if((bankDeck.cards[1].value == "JACK" or bankDeck.cards[2].value == "JACK") and
               (bankDeck.cards[1].value == "TEN" or bankDeck.cards[2].value == "TEN")) then
                return ResultType.LOSEBYJACK
            end
        end

        if(playerValue > 21) then return ResultType.LOSE
        elseif(playerValue == 21) then
            if(bankValue == 21) then return ResultType.TIE
            else return ResultType.WIN
            end
        else
            if(bankValue == 21) then return ResultType.LOSE
            elseif(bankValue > 21) then return ResultType.WIN
            else
                if(bankValue < playerValue) then return ResultType.WIN
                elseif(bankValue == playerValue) then return ResultType.TIE
                else return ResultType.LOSE
                end
            end
        end

    end

    return {
        evaluate = evaluate,
        calculateResult = calculateResult
    }
end

return Calculator